concordium-consensus/src/Concordium/GlobalState/TreeState.hs

```diff
@@ -267,19 +264,27 @@ class (Eq (BlockPointerType m),
     -- as the index of the transaction in the given block).
     -- This will prevent it from being purged while the slot number exceeds
     -- that of the last finalized block.
-    commitTransaction :: Slot -> BlockHash -> BlockItem -> TransactionIndex -> m ()
-    -- |@addCommitTransaction tr slot@ adds a transaction and marks it committed
-    -- for the given slot number. By default the transaction is created in the 'Received' state,
+    commitTransaction :: Slot -> BlockHash -> TVer.BlockItemWithStatus -> TransactionIndex -> m () 
+    -- |@addCommitTransaction ((tr, verResCtx), timestamp) slot@ verifies a transaction within the
+    -- given context and adds the transaction and marks it committed
+    -- for the given slot number. If the transaction was deemed verifiable in the future.
+    -- By default the transaction is created in the 'Received' state,
     -- but if the transaction is already in the table the outcomes are retained.
     -- See documentation of 'AddTransactionResult' for meaning of the return value.
     -- The time is indicative of the receive time of the transaction. It is used to prioritize transactions
     -- when constructing a block.
-    addCommitTransaction :: BlockItem -> Slot -> m AddTransactionResult
-    -- |Purge a transaction from the transaction table if its last committed slot
+    -- Before the transaction is added the transaction will be verified. If it is not ok then it will return
+    -- `NotAdded` together with the verification result.
+    -- The scheduler will need to consult the resulting `VerificationResult` and based on that carry out the correct
+    -- verification before executing the transaction.
+    addCommitTransaction :: (BlockItemWithVerificationContext m, Timestamp) -> Slot -> m AddTransactionResult
+    -- |Purge a transaction from the xbtransaction table if its last committed slot
     -- number does not exceed the slot number of the last finalized block.
     -- (A transaction that has been committed to a finalized block should not be purged.)
     -- Returns @True@ if and only if the transaction is purged.
     purgeTransaction :: BlockItem -> m Bool
+    -- |Zipping a list of transactions with their associated transaction 'VerificationResult'.
+    getTransactionVerificationResults :: [BlockItem] -> m [TVer.BlockItemWithStatus]
```

Comment:
This is currently needed when `addBlock` is being invoked on pending blocks i.e. when we recursively `addBlock` on the children of a just added block.

Reply:
With the latest change this will return (bi, Nothing) when queried with a finalized transaction. It is still used in addBlock such that we avoid duplicate verification computations when possible.